sopel/plugin.py

```diff
@@ -830,31 +833,211 @@ def add_attribute(function):
     return add_attribute
 
 
-def rate(user: int = 0, channel: int = 0, server: int = 0) -> typing.Callable:
+def rate(
+    user: int = 0,
+    channel: int = 0,
+    server: int = 0,
+    *,
+    message: typing.Optional[str] = None,
+) -> typing.Callable:
     """Decorate a function to be rate-limited.
 
-    :param int user: seconds between permitted calls of this function by the
-                     same user
-    :param int channel: seconds between permitted calls of this function in
-                        the same channel, regardless of triggering user
-    :param int server: seconds between permitted calls of this function no
-                       matter who triggered it or where
+    :param user: seconds between permitted calls of this function by the same
+                 user
+    :param channel: seconds between permitted calls of this function in the
+                    same channel, regardless of triggering user
+    :param server: seconds between permitted calls of this function no matter
+                   who triggered it or where
+    :param message: optional keyword argument; default message sent as NOTICE
+                    when a rate limit is reached
 
     How often a function can be triggered on a per-user basis, in a channel,
     or across the server (bot) can be controlled with this decorator. A value
     of ``0`` means no limit. If a function is given a rate of 20, that
     function may only be used once every 20 seconds in the scope corresponding
-    to the parameter. Users on the admin list in Sopel’s configuration are
-    exempted from rate limits.
+    to the parameter::
+
+        from sopel import plugin
+
+        @plugin.rate(10)
+            # won't trigger if used more than once per 10s by a user
+
+        @plugin.rate(10, 10)
+            # won't trigger if used more than once per 10s by a user/channel
+
+        @plugin.rate(10, 10, 2)
+            # won't trigger if used more than once per 10s by a user/channel
+            # and never more than once every 2s
+
+    If a ``message`` is provided, it will be used as the default message sent
+    as a ``NOTICE`` to the user who hit the rate limit::
+
+        @rate(10, 10, 10, message='Hit the rate limit for this function.')
+            # will send a NOTICE
+
+    The message can contain a placeholder for ``nick``: the message will be
+    formatted with the nick that hit rate limit::
+
+        @rate(10, 10, 2, message='Sorry {nick}, you hit the rate limit!')
 
     Rate-limited functions that use scheduled future commands should import
     :class:`threading.Timer` instead of :mod:`sched`, or rate limiting will
     not work properly.
+
+    .. versionchanged:: 8.0
+
+        Optional keyword argument ``message`` was added in Sopel 8.
+
+    .. note::
+
+        Users on the admin list in Sopel’s configuration are exempted from rate
+        limits.
```

Comment:
This Unicode single quote should just be a regular old ASCII `'`.

Yes I know it was already there in the old line you cut/copied, but we do like fixing weird stuff whenever we're already touching it. There are other occurrences elsewhere in this file, so I recommend a global find-and-replace in your IDE/editor.

If you're feeling ambitious you can do a quick search through the other files this touches to find any other instances, but I wouldn't expect any.